### DISCUSSION TITLE
Simply implementing custom value equality

### DISCUSSION THREAD

**Discussion by jaigak** [dotnet/csharplang#7219] [Language Ideas]
Currently in C# implementing value equality for a custom class with the usual expectation requires overriding operator==, operator!=, object.Equals and IEquatable<T>, which is very tedious and also there's no way to make it easier without using source generators. I would naturally expect that operator== maps to object.Equals, but that isn't the case.

Here's an example of a simple IntWrapper class with value equality.
```csharp
class IntWrapper : IEquatable<IntWrapper>
{
  public int Value;
  
  public static bool operator==(IntWrapper? left, IntWrapper? right)
  {
    if (ReferenceEquals(left, right))
    {
      return true;
    }
    
    return left.Value == right.Value;
  }

  public static bool operator!=(IntWrapper? left, IntWrapper? right)
  {
    if (ReferenceEquals(left, right))
    {
      return true;
    }
    
    return !(left == right);   
  }

  public virtual bool Equals(object other)
  {
    if (other is IntWrapper otherWrapper)
    {
      return Value == otherWrapper.Value;
    }
    
    return false;
  }

  bool IEquatable<IntWrapper>.Equals(IntWrapper other)
  {
    return Value == other.Value;
  }
}
```
This is too much code. I'm proposing a non-static version of operator==, which the compiler automatically translates to all these:
```csharp
class IntWrapper
{
    public int Value;
    bool operator==(IntWrapper other)
    {
        // No need to care about nulls or different types, write the value equality operator
        return Value == other.Value;
    }
    // optionally bool operator!=(IntWrapper other), which will be operator== negated if not written
}
```
The generated code should look like the current one you'd write by hand. Object.Equals, operator== (static one that's nullable and actually used by the consuming code), operator!= and IEquatable<T>. It should handle nulls (two nulls equal and a null cannot equal another instance) as well.

**Comment by hez2010** [comment]
This is exactly one of the issues that records have resolved.

**Reply by jaigak** [reply]
No, because there are still differences between a record and a class. Sometimes you do really want a class type with custom equality.

**Reply by DavidArno** [reply]
@JaiganeshKumaran, could you clarify what specific differences you are after between a record and what you are requesting? As hez2010 says, providing a class with built-in value-based equality is one of the primary reasons records exist.

**Reply by jaigak** [reply]
Can you customise the record's value equality?

**Reply by theunrepentantgeek** [reply]
Only by implementing `Equals()` and `GetHashcode()` yourself. 

There's no halfway point - either you use the default provided by the compiler for the record, or you do it yourself.

**Comment by HaloFour** [comment]
This functionality can be accomplished through a source generator where you write the implementation once and the source generator emits the rest of the members.  In a sense that is how records work given that the rest of the members are dispatched through `IEquatable<T>.Equals`.

**Comment by TonyValenti** [comment]
I'll +1 the answer for records. 

Since records were released, our internal design philosophy is this:
Classes act on data. 
Records are data. 

We believe that Classes should only be compared with reference equality.
